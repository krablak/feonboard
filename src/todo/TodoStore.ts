import { Form } from 'src/common/Form';
import { observable, computed, action } from 'mobx';

/**
 * Provides data and actions for TODO app view components.
 */
export class TodoStore {

    /**
     * Form data instance.
     */
    form = new TodoForm();

    /**
     * Todo list items.
     */
    @observable todos: TodoItem[] = [];

    /**
     * Computed property providing flag that exists at least one todo item.
     */
    @computed get hasTodos(): boolean {
        return this.todos.length > 0;
    }

    /**
     * Reads content of form and creates new TODO item.
     */
    @action add() {
        // Check if any text is filled in form field
        if (this.form.fields.newTodo.trim() !== '') {
            // Add new item to TODOs list
            this.todos.unshift({ id: new Date().getTime().toString(), completed: false, title: this.form.fields.newTodo });
            // Clear form field value
            this.form.fields.newTodo = '';
        }
    }

    /**
     * Toggles state of all todo items.
     */
    @action toggleAll = (e: React.ChangeEvent<HTMLInputElement>) => {
        // Use checkbox element 'checked' property to resolve to which state should be item switched
        let newCompletedValue = e.target.checked;
        // Go over all TODO items and change their completet flag        
        this.todos.forEach(item => {
            // Update item completed flag
            item.completed = newCompletedValue;
        });
    }

    /**
     * Changes completed state flag on item.
     */
    @action changeState = (item: TodoItem) => {
        item.completed = !item.completed;
    }

    /**
     * Removes item from TODO list.
     */
    @action remove = (item: TodoItem) => {
        this.todos = this.todos.filter(curItem => curItem.id !== item.id);
    }

}

/**
 * Single TODO item data.
 */
export interface TodoItem {
    // Item unique identifier
    id: string;
    // Item title
    title: string;
    // Completed state flag.
    completed: boolean;
}

/**
 * Holds values of form fields and related onChange action.
 */
export class TodoForm extends Form {

    @observable fields = {
        newTodo: ''
    };

}