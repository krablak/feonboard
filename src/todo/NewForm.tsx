import { TodoStore } from 'src/todo/TodoStore';
import { observer } from 'mobx-react';
import * as React from 'react';

/**
 * Defines props for NewForm UI component.
 */
export interface NewFormProps {
    /**
     * Reference to store providing form data and actions.
     */
    store: TodoStore;
}

/**
 * UI component representing new todo form.
 */
export const NewForm: React.SFC<NewFormProps> = observer((props) => {
    // Shortcut to store form instance
    let form = props.store.form;
    // Event handler of 'enter' key event to add new TODO list item
    const onEnterAdd = (e: React.KeyboardEvent<HTMLInputElement>) => {
        if (e.keyCode === 13) {
            props.store.add();
        }
    };
    return (
        <React.Fragment>
            <input
                // 'name' attribute is used to identifify related form field
                name="newTodo"
                // 'value' attribute is reference to form field value
                value={form.fields.newTodo}
                // 'onChange' attribute is connected to form action which updates form field value found by input element name
                onChange={form.onChange}
                // 'onKeyDown' is connected to store action creating new TODO item
                onKeyDown={onEnterAdd}
                className="new-todo" placeholder="What needs to be done?" />
            {/* Show toggle button only when TODO list is not empty */}
            {props.store.hasTodos &&
                <React.Fragment>
                    <input
                        id="toggle-all"
                        // 'onChange' event is handled by custom store action which updates state of all TODO items.
                        onChange={props.store.toggleAll}
                        className="toggle-all" type="checkbox" />
                    <label htmlFor="toggle-all" />
                </React.Fragment>
            }
        </React.Fragment>
    );
});