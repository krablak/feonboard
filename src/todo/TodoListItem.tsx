import { observer } from 'mobx-react';
import * as React from 'react';
import { TodoItem } from 'src/todo/TodoStore';

/**
 * Defines props for TodoList UI component.
 */
export interface TodoListItemProps {
    // Reference to TODO list item UI data
    item: TodoItem;
    // Function executed on 'Complete item' button click
    onStateChange: (item: TodoItem) => void;
    // Function executed when remove button is clicked
    onRemove: (item: TodoItem) => void;
}

/**
 * UI component representing single TODO item.
 */
export const TodoListItem: React.SFC<TodoListItemProps> = observer((props: TodoListItemProps) => {
    let item = props.item;
    return (
        <li className={item.completed ? 'completed' : ''}>
            <div className="view">
                <input
                    // Input will be checked according to given item completed value
                    checked={item.completed}
                    // Register on state change function from props to checkbox change event
                    onChange={e => props.onStateChange(item)}
                    className="toggle" type="checkbox"
                />
                <label>{item.title}</label>
                <button
                    // Register on item remove function from props to button click event
                    onClick={e => props.onRemove(item)}
                    className="destroy"
                />
            </div>
        </li>
    );
});