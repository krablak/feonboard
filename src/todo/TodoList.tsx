import { observer } from 'mobx-react';
import * as React from 'react';
import { TodoStore } from 'src/todo/TodoStore';
import { TodoListItem } from 'src/todo/TodoListItem';

/**
 * Defines props for TodoList UI component.
 */
export interface TodoListProps {
    /**
     * Reference to store providing form data and actions.
     */
    store: TodoStore;
}

/**
 * UI component representing list of existing TODO list items.
 */
export const TodoList: React.SFC<TodoListProps> = observer((props: TodoListProps) => {
    // Simple shortcut variable to access all existing items
    let todos = props.store.todos;
    return (
        <ul className="todo-list">
            {todos.map(item => (
                <TodoListItem
                    // Why key? https://reactjs.org/docs/lists-and-keys.html
                    key={`todo_item_${item.id}`}
                    // Item data to be rendered
                    item={item}
                    // Register item handler functions to store actions
                    onStateChange={props.store.changeState}
                    onRemove={props.store.remove}
                />
            ))}
        </ul>
    );
});