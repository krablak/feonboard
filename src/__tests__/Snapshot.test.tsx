import * as React from 'react';
import { TodoListItem } from 'src/todo/TodoListItem';
const renderer = require('react-test-renderer');

it('TodoListItem snapshot test', () => {
    const todoListItem0 = renderer.create(<TodoListItem item={{ id: '0', completed: false, title: 'title' }} onRemove={item => false} onStateChange={item => false} />).toJSON();
    expect(todoListItem0).toMatchSnapshot();
});