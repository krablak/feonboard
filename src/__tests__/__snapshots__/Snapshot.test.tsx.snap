// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoListItem snapshot test 1`] = `
<li
  className=""
>
  <div
    className="view"
  >
    <input
      checked={false}
      className="toggle"
      onChange={[Function]}
      type="checkbox"
    />
    <label>
      title
    </label>
    <button
      className="destroy"
      onClick={[Function]}
    />
  </div>
</li>
`;
